app/forms/change_note_form.rb

```diff
@@ -0,0 +1,39 @@
+class ChangeNoteForm
+  include ActiveModel::Model
+
+  attr_accessor :minor_change, :change_note
+
+  validates :minor_change, inclusion: { in: [true, false] }
+  validates :change_note, presence: true, if: :major_version?
+
+  def self.build_from_edition(edition)
+    new(
+      minor_change: edition.minor_change,
+      change_note: edition.change_note,
+    )
+  end
+
+  def save!(edition)
+    self.minor_change = ActiveModel::Type::Boolean.new.cast(minor_change)
```

Comment:
this feels a bit clunky. It's only really needed if the save is unsuccessful because Rails is smart enough to do cast on save.

```
object.attribute
=> true

object.update(attribute: "false")

object.attribute
=> false
```
It needs be cast again before the page is re-rendered though, as the radios treat any string as true. Without the minor radio will always be selected.

It could potentially live somewhere else 


```
if invalid?
  self.minor_change = ActiveModel::Type::Boolean.new.cast(minor_change)
  return false
elsif major_version?
else 
end 
```

etc.
Or it could be a public method called in the controller. It might be fine as is though. Whichever way is slightly annoying but i don't think it's a massive deal.

Keen to hear peoples thoughts.

Reply:
Would including`ActiveRecord::AttributeAssignment` on the form model help with this? I used it [here](https://github.com/alphagov/whitehall/blob/44cefdf037a58d32f68f21577da5a725599fefbb/test/unit/app/models/date_validation_test.rb#L11C4-L11C46) to get Rails to do its magic casting in a test

Reply:
I could def be wrong but i suspect it already has access to `assign_attributes` as it inherits from active model.

I just tested using `assign_attributes` locally and it doesn't seem to cast booleans. I might well be being dense though! Waiting for the coffee to kick in 😂 maybe we can discuss after standup?

Reply:
Apologies for being nosey again, I was intrigued by the PR title

You could maybe use the `attribute` method like we've done here https://github.com/alphagov/transition/blob/ac1217bd251d42b8d3618aa93f825fb9fc8fe9e8/app/forms/site_form.rb#L15

Reply:
I think you probably want to be using [ActiveModel::Attributes](https://rubydoc.info/github/rails/rails/main/ActiveModel/Attributes) for this rather than [AttributeAssignment](https://rubydoc.info/github/rails/rails/main/ActiveModel/AttributeAssignment).

> Active Model Attributes are aware of data types, can have default values, and can handle casting and serialization.

So you could have something like:

```ruby
class ChangeNoteForm
  include ActiveModel::Model
  include ActiveModel::Attributes

  attribute :minor_change, :boolean, default: false
end
```

and that should handle casting the value automatically.

Reply:
Thanks! Have sorted it